### ISSUE TITLE
Possible memory leak of grpc_ssl_credentials (grpc_call leak if response stream not consumed by client)

### ISSUE THREAD

**Issue by tomes92** [grpc/grpc#28153]
### What version of gRPC and what language are you using?

gRPC C# v2.42.0

### What operating system (Linux, Windows,...) and version?

Windows 10 Pro 19043 x64

### What did you do?

Disposal of a gRPC client , created with a root certificates string, doesn't seem to free this string in unmanaged memory. The code below repeatedly creates and disposes a gRPC client. No running gRPC service is required and a random root certificates string can be used to reproduce this memory leak on my system.

Proto file:

```proto
syntax = "proto3";

package communication;

service MessageService {
	rpc MessagePipeline(stream Message) returns (stream Message);
}

message Message {
}
```

C# code:

```csharp
using Communication;
using Grpc.Core;
using System;
using System.Linq;
using System.Threading;

namespace Client
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var client = new MessageClient())
            {
                client.Start();

                Console.ReadKey();
            }
        }
    }

    class MessageClient : IDisposable
    {
        const int reconnectDelay = 1000; //ms

        bool bDisposed;

        Thread connectThread;
        Channel channel;

        public void Dispose()
        {
            if (bDisposed)
                return;

            bDisposed = true;

            if (connectThread != null)
            {
                if (connectThread.ThreadState != ThreadState.Unstarted)
                    connectThread.Join();

                connectThread = null;
            }
        }

        public bool Start()
        {
            connectThread = new Thread(new ThreadStart(RunConnect));
            connectThread.Name = "Connect thread";
            connectThread.Start();

            return true;
        }

        #region Connection

        /*internal static string GetCertificatePem(X509Certificate2 cert)
        {
            var builder = new StringBuilder();
            builder.AppendLine("-----BEGIN CERTIFICATE-----");
            builder.AppendLine(Convert.ToBase64String(cert.RawData, Base64FormattingOptions.InsertLineBreaks));
            builder.AppendLine("-----END CERTIFICATE-----");

            return builder.ToString();
        }

        internal static string GetRootCertificates()
        {
            StringBuilder builder = new StringBuilder();

            using (var store = new X509Store(StoreName.Root))
            {
                store.Open(OpenFlags.ReadOnly);

                foreach (X509Certificate2 mCert in store.Certificates)
                {
                    builder.AppendLine(
                        "# Issuer: " + mCert.Issuer.ToString() + "\n" +
                        "# Subject: " + mCert.Subject.ToString() + "\n" +
                        "# Label: " + mCert.FriendlyName.ToString() + "\n" +
                        "# Serial: " + mCert.SerialNumber.ToString() + "\n" +
                        "# SHA1 Fingerprint: " + mCert.GetCertHashString().ToString() + "\n" +
                        GetCertificatePem(mCert) + "\n");
                }
            }

            return builder.ToString();
        }*/

        // Random string is good enough to reproduce the leak
        internal static string GetRootCertificates()
        {
            return new string(Enumerable.Repeat('a', 150000).ToArray());
        }

        // Continuously create and dispose gRPC client
        void RunConnect()
        {
            while (!bDisposed)
            {
                // Create channel
                try
                {
                    string rootCerts = GetRootCertificates();
                    channel = new Channel("localhost", 10000, new SslCredentials(rootCerts));
                }
                catch
                {
                    Thread.Sleep(reconnectDelay);
                    continue;
                }

                // Create client
                var client = new MessageService.MessageServiceClient(channel);
                var connectCall = client.MessagePipeline();

                // Cleanup
                connectCall.Dispose();
                channel.ShutdownAsync().Wait();

                if (!bDisposed)
                    Thread.Sleep(reconnectDelay);
            }
        }
        #endregion
    }
}
```

### What did you expect to see?

Roughly constant memory usage.

### What did you see instead?

Increasing memory usage every time a client is created. The ref counts on the native grpc_ssl_credentials instances never seem to reach 0, even after disposal of the client.

**Comment by tomes92** [comment]
Hi, is there by any chance an update on this?

**Comment by tonydnewell** [comment]
I think the memory leak is in the use of bidirectional streaming rather than SSL credentials.
Testing with a simple non-streaming call does not show a memory leak. Also if you drain the responses (or at least attempt to) then there does not appear to be a memory leak. See the code below.

If you just do this (the original code):

```
var connectCall = client.MessagePipeline();

// Cleanup
connectCall.Dispose();
	
```

then subchannels are created that are never cleaned up properly.

If however you do this then there does not appear to be a memory leak:

```
using (var connectCall = client.MessagePipeline())
{
	// do some work...
	
	// (optional) send message
	try
	{
		await connectCall.RequestStream.WriteAsync(new Message());
	}
	catch (Exception e)
	{
		// handle exception
		Console.WriteLine("##### Exception writing to request stream {0} ", e.Message);
	}

	// read all responses
	// Note: *Must* drain all the responses before abandoning the call
	try
	{
		while (await connectCall.ResponseStream.MoveNext())
		{
			//do something
		}
	}
	catch (Exception e)
	{
		// handle exception
		Console.WriteLine("##### Exception draining response stream {0} ", e.Message);
	}
}

```

**Comment by jtattermusch** [comment]
Related: https://github.com/grpc/grpc/issues/8451#issuecomment-1147943234

**Comment by tonydnewell** [comment]
Further investigation - when the server is unavailable and the client creates a server streaming call (e.g. a duplex call) the connection to the server can not be established. Internally the `OnReceivedResponseHeaders()` callback is called with `success == false`. This 'false' case was never handled so the call was left in a state where the resources where never released.

Created PR with a fix: https://github.com/grpc/grpc/pull/30141

**Comment by tonydnewell** [comment]
New fix PR: https://github.com/grpc/grpc/pull/30290

**Comment by tonydnewell** [comment]
@jtattermusch can close - the fix is in release v1.46.5
